### PR TITLE
Collateralization Oracle Wrapper Refactor

### PR DESCRIPTION
**Pull request by Joeysantoro**
Refactors the CollateralizationOracleWrapper to use more transparent logic around deviation thresholds being exceeded. Also includes an interface and new update method that reverts unless the oracle is currently out of date.

The new update method is used in a keeper that incentivizes updates of the oracle.

### CODE REVIEW

File: contracts/mock/MockCollateralizationOracle.sol [hunk 1]
```diff
@@ -19,10 +19,6 @@ contract MockCollateralizationOracle is MockOracleCoreRef {
         pcvValue = _pcvValue;
     }
 
-    function pcvStats() public view returns (uint256, uint256, int256, bool) {
```
Comment (Joeysantoro) [inline contracts/mock/MockCollateralizationOracle.sol:22]: unrelated cleanup from merge conflict

File: contracts/keeper/CollateralizationOracleKeeper.sol [hunk 2]
```diff
@@ -0,0 +1,33 @@
+pragma solidity ^0.8.0;
+
+import "../token/FeiTimedMinter.sol";
+import "../oracle/ICollateralizationOracleWrapper.sol";
+
+/// @title CollateralizationOracleKeeper
+/// @notice a FEI timed minter which only rewards when updating the collateralization oracle 
+contract CollateralizationOracleKeeper is FeiTimedMinter {
```
Comment (Joeysantoro) [inline contracts/keeper/CollateralizationOracleKeeper.sol:8]: Did not add a unit test for this class because it is such as simple override of the FeiTimedMinter.

File: contracts/oracle/CollateralizationOracleWrapper.sol [hunk 3]
```diff
@@ -177,24 +198,14 @@ contract CollateralizationOracleWrapper is Timed, ICollateralizationOracle, Core
 
         require(_validityStatus, "CollateralizationOracleWrapper: CollateralizationOracle reading is invalid");
 
-        Decimal.D256 memory _thresholdLow = Decimal.from(10_000 - deviationThresholdBasisPoints).div(10_000);
-        Decimal.D256 memory _thresholdHigh = Decimal.from(10_000 + deviationThresholdBasisPoints).div(10_000);
-
-        obsolete = paused();
```
Comment (Joeysantoro) [inline contracts/oracle/CollateralizationOracleWrapper.sol:183]: removed the "paused" condition for exceeded deviation threshold. Feels unnatural to include that as a criteria

File: contracts/oracle/CollateralizationOracleWrapper.sol [hunk 4]
```diff
@@ -110,13 +112,28 @@ contract CollateralizationOracleWrapper is Timed, ICollateralizationOracle, Core
...
     function update() external override whenNotPaused {
+        _update();
+    }
+
+    /** 
+        @notice this method reverts if the oracle is not outdated
+        It is useful if the caller is incentivized for calling only when the deviation threshold or frequency has passed
+    */ 
+    function updateIfOutdated() external override whenNotPaused {
+        require(_update(), "CollateralizationOracleWrapper: outdated");
```
Comment (eswak) [inline contracts/oracle/CollateralizationOracleWrapper.sol:130]: the exception message would rather be "not outdated"

File: contracts/oracle/CollateralizationOracleWrapper.sol [hunk 5]
```diff
@@ -177,24 +198,14 @@ contract CollateralizationOracleWrapper is Timed, ICollateralizationOracle, Core
...
+            || _isExceededDeviationThreshold(cachedUserCirculatingFei, _userCirculatingFei);
+    }
 
-        // Circulating Fei checks
-        Decimal.D256 memory circFeiRatio = Decimal.from(_userCirculatingFei).div(cachedUserCirculatingFei);
-        // CircFEI < threshold
-        obsolete = obsolete || circFeiRatio.lessThan(_thresholdLow);
-        // CircFEI > threshold
-        obsolete = obsolete || circFeiRatio.greaterThan(_thresholdHigh);
+    function _isExceededDeviationThreshold(uint256 cached, uint256 current) internal view returns (bool) {
```
Comment (eswak) [inline contracts/oracle/CollateralizationOracleWrapper.sol:205]: I admit it's more elegant like this 😇

File: contracts/oracle/CollateralizationOracleWrapper.sol (hunk 4 above)
Comment (Joeysantoro) [reply contracts/oracle/CollateralizationOracleWrapper.sol]: Fixed

File: contracts/keeper/CollateralizationOracleKeeper.sol [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        address _core,
+        uint256 _incentive,
+        ICollateralizationOracleWrapper _collateralizationOracleWrapper
+    ) 
+        FeiTimedMinter(_core, address(this), _incentive, MIN_MINT_FREQUENCY, 0) 
+    {
+        collateralizationOracleWrapper = _collateralizationOracleWrapper;
+    }
+
+    function _afterMint() internal override {
```
Comment (Joeysantoro) [inline contracts/keeper/CollateralizationOracleKeeper.sol:29]: Updated to use afterMint hook

File: contracts/oracle/CollateralizationOracleWrapper.sol (hunk 5 above)
Comment (ditchfieldcaleb) [reply contracts/oracle/CollateralizationOracleWrapper.sol:205]: beautiful

**Review by ditchfieldcaleb — APPROVED** [review]
